pkg/scheduler/cache/cache.go

```diff
@@ -176,6 +184,251 @@ type imageState struct {
+	hnfb := sc.hypernodeForestbuilder
+	hntn := NewHyperNodeTreeNode(hn.Name)
+	hntn.HyperNodeRef = hn
+	hnfb.nodes[hn.Name] = hntn
+
+	tierStr := hn.Spec.Tier
+	tier, err := strconv.Atoi(tierStr)
+	if err != nil {
+		klog.Errorf("Failed to convert tier %s to int: %v", tierStr, err)
+		return
+	}
+	hntn.Tier = tier
+	klog.V(4).Infof("Adding hypernode %s to the forest, in %d tier", hn.Name, tier)
+
+	// Iterate over members and add them to the forest
+	for _, member := range hn.Spec.Members {
+		switch member.Type {
+		case networktopov1alpha1.MemberTypeNode:
+			switch member.Selector.Type {
+			case networktopov1alpha1.ExactMatchMemberSelectorType:
+				nodeName := member.Selector.ExactMatch.Name
+				hntn.NodeMatchers.Insert(NewExactMatcher(nodeName))
+				workerNode := sc.Nodes[nodeName]
+				if workerNode == nil {
+					klog.Errorf("WorkerNode %s not found in the cache", nodeName)
+					return
+				}
+				hntn.Nodes.Insert(workerNode)
+			case networktopov1alpha1.RegexMatchMemberSelectorType:
+				matcher, err := NewRegexMatcher(member.Selector.RegexMatch.Pattern)
+				if err != nil {
+					klog.Errorf("Failed to create regex matcher for %s: %v", member.Selector.RegexMatch.Pattern, err)
+					continue
+				}
+				hntn.NodeMatchers.Insert(matcher)
+				for _, nodeInfo := range sc.Nodes {
+					if matcher.Matches(nodeInfo.Name) {
+						hntn.Nodes.Insert(nodeInfo)
+					}
+				}
+			}
+		case networktopov1alpha1.MemberTypeHyperNode:
+			switch member.Selector.Type {
+			case networktopov1alpha1.ExactMatchMemberSelectorType:
+				memberName := member.Selector.ExactMatch.Name
+				hntn.MemberMatchers.Insert(NewExactMatcher(memberName))
+				member := sc.hypernodeForestbuilder.nodes[memberName]
+				if member != nil {
+					insertHyperNodeMember(hntn, hnfb, memberName, member, sc)
+					if hntn.Parent != nil {
+						sc.HyperNodes[hntn.Parent.Name].Union(sc.HyperNodes[memberName])
+					}
+				}
+			case networktopov1alpha1.RegexMatchMemberSelectorType:
+				// TODO
+				matcher, err := NewRegexMatcher(member.Selector.RegexMatch.Pattern)
+				if err != nil {
+					klog.Errorf("Failed to create regex matcher for %s: %v", member.Selector.RegexMatch.Pattern, err)
+					continue
+				}
+				hntn.MemberMatchers.Insert(matcher)
+				for memberName, memberNode := range sc.hypernodeForestbuilder.nodes {
+					if matcher.Matches(memberName) {
+						insertHyperNodeMember(hntn, hnfb, memberName, memberNode, sc)
+
+					}
+				}
+				if hntn.Parent != nil {
+					sc.HyperNodes[hntn.Parent.Name].Union(sc.HyperNodes[hntn.Name])
+				}
+			}
+		}
+	}
+
+	// Iterate over existing HyperNodes, if the name of the new HyperNode matches
+	// either the Exact or Regex matchers of a parent, add it as a child member
+	for _, parentNode := range sc.hypernodeForestbuilder.nodes {
+		for m := range parentNode.MemberMatchers {
+			if m.Exact != "" && m.Exact == hn.Name {
+				insertHyperNodeMember(parentNode, hnfb, hn.Name, hntn, sc)
```

Comment:
insertHyperNodeMember should be used as a method of *HyperNodeForestBuilder instead of passing it in through parameters.

Reply:
SchedulerCache should also be used as an attribute of HyperNodeForestBuilder. The function parameters must be consistent with the function name and responsibilities.

Reply:
Don't memerNode already contain a name attribute? No need to pass again